Support paged cgm

tmecklem opened the pull request:
Summary: Add support for fetching and displaying recent data from cgm pages. See #200 

This PR adds to MinimedKit the ability to fetch and decode cgm pages into events similar to the way that history pages are processed. It also supports fetching the pointer to the current page of cgm data to support fetching the most recent pages first.

In RileyLinkKit, functionality is built upon the additions to MinimedKit to fetch recent cgm data from the last 24 hours and display it similarly to the fetch recent history functionality.

tmecklem committed:
Add import Foundation to HistoryPage

tmecklem committed:
Add GlucosePage and enumeration of event types

Add GlucosePage CRC check tests and Data extension to reverseBytes since
GlucosePages are easier to decode in reverse.

tmecklem committed:
Add reverseBytes extension to Data

tmecklem committed:
Ensure that event name sequence matches decocare output

In writing this test and the code that passed the test, I encountered a
strange mismatch from decocare. Decocare had more sgv entries, but they
had values of 0. Additional discovery and discussion led to a change in
decocare: https://github.com/openaps/decocare/pull/11

tmecklem committed:
Decode independent events

Independent glucose events are events that have a timestamp in the 
record and do not serve as a reference timestamp for other records.

tmecklem committed:
Add event parsing for reference timestamp glucose records

Reference timestamped records are used to apply a timestamp to other 
records using an offset (5 minutes).

tmecklem committed:
Add event parsing for relative timestamp records

Relative timestamp records need a timestamp, but don’t have one in their
own raw data, so that’s not included in this commit.

tmecklem committed:
Add relative timestamps to records when parsing page data

tmecklem committed:
Add MessageBody classes for fetching glucose pages

tmecklem committed:
Add relevant data to dictionary representations

tmecklem committed:
Add handling for currently unknown opcodes

tmecklem committed:
Add pump ops for fetching recent glucose

tmecklem committed:
Fix incorrect matching of unknown glucose events

Clean up test output a little to help in troubleshooting

ps2 commented on MinimedKit/Extensions/NSData.swift:
```diff
@@ -71,6 +71,14 @@ extension Data {
 }
 
+extension Data {
+    func reverseBytes() -> Data {
```
There is a method in the standard library to do this already.  See https://developer.apple.com/reference/foundation/data/1780245-reversed

tmecklem replied:
I'm having trouble figuring out what to do with the data structure that reversed() gives me. It's a `ReversedRandomAccessCollection<Data>`, and it looks like I have to use a different Index structure. The change seems to leak all the way through to how I access rawData GlucoseEvent and seems messy. Can you take a look and help me understand how to appropriately use reversed() on the page Data?

ps2 replied:
You can use Data(data.reversed()) to create a new reversed data object.  There may be ways to work with RandomAccessCollection<Data> types, and avoid the copy, but I wasn't able to get that working in my swift playground.

tmecklem replied:
Thanks! Change made.

ps2 commented on RileyLinkKit/PumpOpsSynchronous.swift:
```diff
@@ -544,6 +544,106 @@ class PumpOpsSynchronous {
+        var events = [TimestampedGlucoseEvent]()
+        
+        // Going to scan backwards in time through events, so event time should be monotonically decreasing.
+        let eventTimestampDeltaAllowance = TimeInterval(hours: 9)
```
This should not be necessary, unless you have encountered out of order glucose events, and need to keep looking back further in history to ensure you get all events since startDate.

tmecklem replied:
There are out of order events due to relative timestamping. I changed the timestamp check on https://github.com/ps2/rileylink_ios/pull/270/commits/f266ebbf41e9888470d9db92285018331b2afda4#diff-bf2d04fbde756ea81dd46dafafbdacddR589 to use the reference timestamped events only. That should ensure that all relevant glucose events are captured since startDate at the expense of maybe capturing a few extra independently timestamped (non-sgv) entries.

ps2 commented on RileyLinkKit/PumpOpsSynchronous.swift:
```diff
@@ -544,6 +544,106 @@ class PumpOpsSynchronous {
+                idx = top
+            }
+            
+            let page = try GlucosePage(pageData: pageData, pumpModel: pumpModel)
```
I searched and didn't see any actual dependency on pumpModel in the data parsing.  Can this be removed?

tmecklem replied:
Good catch! 👍

ps2 reviewed (changes requested):
Amazing work, Tim! Looking forward to getting this into Loop.  It would also be nice to add functionality for CGM upload to NS for those users using just the RL app, though I think that's a smaller set.

tmecklem committed:
Remove unnecessary passing of pumpModel

tmecklem committed:
Remove timestamp delta allowance and use reference timestamps in check

Glucose events can be out of order (and are frequently). Checking for
only reference timestamps should catch all relevant events.

tmecklem committed:
Skip UnknownGlucoseEvents rather than try to timestamp them

ps2 commented on RileyLinkKit/PumpOps.swift:
```diff
@@ -144,6 +144,34 @@ public class PumpOps {
+     - failure(error):  An error describing why the command failed
+     
+     */
+    public func getGlucoseHistoryEvents(since startDate: Date, completion: @escaping (Either<(events: [TimestampedGlucoseEvent], pumpModel: PumpModel), Error>) -> Void) {
```
Does pumpModel need to be here?

tmecklem replied:
Nope!

tmecklem committed:
Use reversed() instead of custom reverseBytes()

tmecklem commented:
I'm planning to get this version of the app installed on the phone of someone using a 722 pump with enlite before the start of the weekend for Real World:tm: testing.

tmecklem committed:
Remove pumpModel from PumpOps
